src/bindgen/bindings.rs

```diff
@@ -367,19 +247,19 @@ impl Bindings {
             match *item {
                 ItemContainer::Constant(..) => unreachable!(),
                 ItemContainer::Static(..) => unreachable!(),
-                ItemContainer::Enum(ref x) => x.write(&self.config, &mut out),
-                ItemContainer::Struct(ref x) => x.write(&self.config, &mut out),
-                ItemContainer::Union(ref x) => x.write(&self.config, &mut out),
-                ItemContainer::OpaqueItem(ref x) => x.write(&self.config, &mut out),
-                ItemContainer::Typedef(ref x) => x.write(&self.config, &mut out),
+                ItemContainer::Enum(ref x) => language_backend.write_enum(&mut out, x),
+                ItemContainer::Struct(ref x) => language_backend.write_struct(&mut out, x),
+                ItemContainer::Union(ref x) => language_backend.write_union(&mut out, x),
+                ItemContainer::OpaqueItem(ref x) => language_backend.write_opaque_item(&mut out, x),
+                ItemContainer::Typedef(ref x) => language_backend.write_type_def(&mut out, x),
             }
```

Comment:
I'm not too sure about the way you've abstracted the language backend. The problem is that each item is passed one at a time to the language backend. This works fine for simple languages where you can write any of these items anywhere and they only need to be in one place, but will cause problems for more complicated setups.

Consider a hypothetical language where in a source file, all the functions have to be defined before the structs. This abstraction would make writing a backend for that language very annoying, because the structs are always written first. Or, consider the situation that is making me write this in the first place; in my C# bindings for performance and flexibility of the generated file, I want to have a struct which contains all the function pointers dynamically loaded from the shared library. The problem is that in addition to the struct with the raw function pointers, I want to have nice and pretty C# methods which call the lower level function pointers (consumers of a native library in C# won't want to be dealing with function pointers). The way this code stands, this is unnecessarily difficult to do because each function needs to be in two places in the generated file, once as a field in the function pointer struct and second the method consumers will be using, but the abstraction expects me to just be able to write any function right after all the type definitions.

In my opinion, basically this whole `write_with_backend` function should be a part of the backend itself. Ideally, the backend would get passed a source writer, the configs and a struct containing a bunch of lists of all the enums, structs, unions, opaque items, type definitions, constants, globals and functions. Then, it's up to the backend to loop through these lists as is it pleases and can reason about the bindings as a whole rather than being drip-fed them one item at a time.

Reply:
Having this code implemented inside the backend would mean :
  - duplicated code for most languages
  - harder implementation of new languages: right now implement the trait and you've a function that is called for every item that probably needs your attention this won't be the case if we move this reponsability t the backend

For your use case, what I would do is save the items you want to use and actually do the write in either `open_close_namespace(Close)`  or `write_footer()`. I can make the trait functions to take a `&mut self` instead of a `&self` in order to prevent having to have a `Mutex` to do that. Would that be a solution for you ?

Reply:
Yes, that is a solution. I would still maintain that the order of things within the output, and things like the astetic newlines this method is writing between the different types of things should be a responsibility of the backend. I believe I'd be nicer to keep the backends stateless, avoid copying data unnecessarily and avoid the weirdness of writing function data in the function which is supposed to be writing the footer, especially considering the price for all that flexibility is duplicating code to loop over the data. But, if you don't share that opinion than I will happily implement your suggested solution.

Reply:
I have found a solution that is in fact way better: I just pushed changes that indeed creates a `write_bindings`  function in the `LanguageBackend` trait, but provides a default implementation for it. This way most languages should still only have to implement the abstract functions in the trait and have something that works but we can tweak the default in case  we need it. This allowed me to remove a few of the remaning `if`s testing agaist the language in the generic impl \o/. 

All functions in the trait now also take `&mut self` instead of `&self` this should help impl for languages that need to remember stuff between calls